contracts/test/spaces/membership/unit/MembershipJoinSpace.t.sol

```diff
@@ -76,6 +78,68 @@ contract MembershipJoinSpace is
     address[] selectedNodes;
   }
 
+  function _signMessage(
+    uint256 privateKey,
+    bytes32 message
+  ) internal pure returns (bytes memory) {
+    (uint8 v, bytes32 r, bytes32 s) = vm.sign(
+      privateKey,
+      MessageHashUtils.toEthSignedMessageHash(message)
+    );
+    return abi.encodePacked(r, s, v);
+  }
+
+  function test_joinSpaceWithUserEntitlement_passes() external {
+    vm.prank(alice);
+    membership.joinSpace(alice);
+    assertEq(membership.balanceOf(alice), 1);
+  }
+
+  function test_joinSpaceWithRootWalletUserEntitlement_passes() external {
+    IWalletLink wl = IWalletLink(spaceFactory);
+    Vm.Wallet memory daveWallet = vm.createWallet("dave");
+
+    uint256 nonce = walletLink.getLatestNonceForRootKey(daveWallet.addr);
+    bytes32 messageHash = keccak256(abi.encode(daveWallet.addr, nonce));
+    bytes memory signature = _signMessage(aliceWallet.privateKey, messageHash);
+
+    vm.startPrank(daveWallet.addr);
+    vm.expectEmit(address(wl));
+    emit IWalletLinkBase.LinkWalletToRootKey(daveWallet.addr, aliceWallet.addr);
+    walletLink.linkCallerToRootKey(
+      IWalletLinkBase.LinkedWallet(aliceWallet.addr, signature),
+      nonce
+    );
+
+    membership.joinSpace(daveWallet.addr);
+    assertEq(membership.balanceOf(daveWallet.addr), 1);
+  }
+
+  function test_joinSpaceWithLinkedWalletUserEntitlement_passes() external {
+    IWalletLink wl = IWalletLink(spaceFactory);
+    Vm.Wallet memory emilyWallet = vm.createWallet("emily");
+
+    uint256 nonce = walletLink.getLatestNonceForRootKey(aliceWallet.addr);
+    bytes32 messageHash = keccak256(abi.encode(aliceWallet.addr, nonce));
+    bytes memory signature = _signMessage(emilyWallet.privateKey, messageHash);
+
+    vm.startPrank(alice);
+    vm.expectEmit(address(wl));
+    emit IWalletLinkBase.LinkWalletToRootKey(
+      aliceWallet.addr,
+      emilyWallet.addr
+    );
+    walletLink.linkCallerToRootKey(
+      IWalletLinkBase.LinkedWallet(emilyWallet.addr, signature),
```

Comment:
import IWalletLinkBase into test contract to avoid using IWalletLinkBase.LinkedWallet